File: src/lib/supportedFiles.ts
/**
 * File formats the editor can open as text. Image assets are listed separately
 * (`IMAGE_ASSET_EXTENSIONS` / `isImageAsset`) for the docs tree and preview
 * tabs — they must not be added to `SUPPORTED_EXTENSIONS` (read/write/index).
 */
import { extensionOf } from "./fileExtensions";
import { ASCIIDOC_LANGUAGE_ID } from "../monaco/asciidocLanguage";

export const SUPPORTED_EXTENSIONS = [
  ".adoc",
  ".asciidoc",
  ".json",
  ".md",
  ".markdown",
  ".txt",
  ".puml",
  ".plantuml",
  ".yaml",
  ".yml",
  ".mmd",
  ".mermaid",
] as const;

/** Keep in sync with `IMAGE_ASSET_EXTENSIONS` in `src-tauri/.../supported_files.rs`. */
export const IMAGE_ASSET_EXTENSIONS = [
  ".png",
  ".jpg",
  ".jpeg",
  ".gif",
  ".svg",
  ".webp",
  ".bmp",
  ".ico",
] as const;

export type SupportedExtension = (typeof SUPPORTED_EXTENSIONS)[number];

/** Human-readable labels for the welcome screen / docs. */
export const SUPPORTED_FORMAT_LABELS = [
  "AsciiDoc (.adoc, .asciidoc)",
  "JSON (.json)",
  "Markdown (.md, .markdown)",
  "Plain text (.txt)",
  "PlantUML (.puml, .plantuml)",
  "YAML (.yaml, .yml)",
  "Mermaid (.mmd, .mermaid)",
] as const;

/** Primary extensions offered when creating a new file (IDEA-style picker). */
export const NEW_FILE_EXTENSION_OPTIONS = [
  { ext: ".adoc", label: "AsciiDoc" },
  { ext: ".json", label: "JSON" },
  { ext: ".md", label: "Markdown" },
  { ext: ".txt", label: "Plain text" },
  { ext: ".puml", label: "PlantUML" },
  { ext: ".yaml", label: "YAML" },
  { ext: ".mmd", label: "Mermaid" },
] as const;

export const DEFAULT_NEW_FILE_EXTENSION = ".adoc" as const;

export function isSupportedFile(path: string): boolean {
  const ext = extensionOf(path);
  return (SUPPORTED_EXTENSIONS as readonly string[]).includes(ext);
}

/** True for image binaries shown in the docs tree / opened as preview tabs. */
export function isImageAsset(path: string): boolean {
  const ext = extensionOf(path);
  return (IMAGE_ASSET_EXTENSIONS as readonly string[]).includes(ext);
}

/** True when the path is an AsciiDoc document (`.adoc` or `.asciidoc`). */
export function isAsciiDocPath(path: string): boolean {
  const ext = extensionOf(path);
  return ext === ".adoc" || ext === ".asciidoc";
}

/** Monaco language id for a supported path; unknown → plaintext. */
export function monacoLanguageFor(path: string): string {
  switch (extensionOf(path)) {
    case ".json":
      return "json";
    case ".md":
    case ".markdown":
      return "markdown";
    case ".yaml":
    case ".yml":
      return "yaml";
    case ".adoc":
    case ".asciidoc":
      return ASCIIDOC_LANGUAGE_ID;
    case ".txt":
    case ".puml":
    case ".plantuml":
    case ".mmd":
    case ".mermaid":
    default:
      return "plaintext";
  }
}

/** Human-readable format label for the status bar. */
export function formatLabelFor(path: string): string {
  if (isImageAsset(path)) return "Image";
  switch (extensionOf(path)) {
    case ".adoc":
    case ".asciidoc":
      return "AsciiDoc";
    case ".json":
      return "JSON";
    case ".md":
    case ".markdown":
      return "Markdown";
    case ".txt":
      return "Plain text";
    case ".puml":
    case ".plantuml":
      return "PlantUML";
    case ".yaml":
    case ".yml":
      return "YAML";
    case ".mmd":
    case ".mermaid":
      return "Mermaid";
    default:
      return "Plain text";
  }
}

/** Detect dominant line ending from file contents for the status bar. */
export function lineEndingLabelFor(content: string): string {
  const crlf = (content.match(/\r\n/g) ?? []).length;
  const crOnly = (content.match(/\r(?!\n)/g) ?? []).length;
  const lfOnly = (content.match(/(?<!\r)\n/g) ?? []).length;

  if (crlf === 0 && crOnly === 0 && lfOnly === 0) return "LF";
  if (crlf >= lfOnly && crlf >= crOnly) return "CRLF";
  if (crOnly > lfOnly) return "CR";
  return "LF";
}

File: src-tauri/src/domain/supported_files.rs
/// Extensions the editor can open. Keep in sync with `src/lib/supportedFiles.ts`.
pub const SUPPORTED_EXTENSIONS: &[&str] = &[
    ".adoc",
    ".asciidoc",
    ".json",
    ".md",
    ".markdown",
    ".txt",
    ".puml",
    ".plantuml",
    ".yaml",
    ".yml",
    ".mmd",
    ".mermaid",
];

/// Image asset extensions discoverable under docsRoot for `image::` completions.
pub const IMAGE_ASSET_EXTENSIONS: &[&str] = &[
    ".png", ".jpg", ".jpeg", ".gif", ".svg", ".webp", ".bmp", ".ico",
];

pub fn extension_of(path: &str) -> String {
    let base = path.rsplit(['/', '\\']).next().unwrap_or(path);
    let Some(dot) = base.rfind('.') else {
        return String::new();
    };
    if dot == 0 {
        return String::new();
    }
    base[dot..].to_ascii_lowercase()
}

pub fn is_supported_file(path: &str) -> bool {
    let ext = extension_of(path);
    SUPPORTED_EXTENSIONS.contains(&ext.as_str())
}

pub fn is_image_asset(path: &str) -> bool {
    let ext = extension_of(path);
    IMAGE_ASSET_EXTENSIONS.contains(&ext.as_str())
}

/// Docs-tree / rename-copy allowlist: editable docs plus image assets.
/// Does **not** replace `is_supported_file` for read/write/index.
pub fn is_docs_tree_file(path: &str) -> bool {
    is_supported_file(path) || is_image_asset(path)
}

pub fn is_asciidoc(path: &str) -> bool {
    matches!(extension_of(path).as_str(), ".adoc" | ".asciidoc")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detects_supported_extensions() {
        assert!(is_supported_file("foo/bar.adoc"));
        assert!(is_supported_file(r"C:\docs\x.JSON"));
        assert!(!is_supported_file("main.rs"));
        assert!(!is_supported_file(".gitignore"));
    }

    #[test]
    fn asciidoc_helpers() {
        assert!(is_asciidoc("a.adoc"));
        assert!(!is_asciidoc("a.md"));
    }

    #[test]
    fn detects_image_assets() {
        assert!(is_image_asset("images/logo.PNG"));
        assert!(is_image_asset("a.svg"));
        assert!(!is_image_asset("a.adoc"));
        assert!(!is_image_asset("a.rs"));
    }

    #[test]
    fn docs_tree_file_includes_images_but_supported_does_not() {
        assert!(is_docs_tree_file("a.png"));
        assert!(is_docs_tree_file("a.adoc"));
        assert!(!is_supported_file("a.png"));
    }
}
